fvm/transactionInvoker.go

```diff
@@ -45,129 +44,29 @@ func (i TransactionInvoker) Process(
 	}
 
 	var modifiedSets programsCache.ModifiedSetsInvalidator
-	defer func() {
-		if errors.IsFailure(processErr) {
-			// Just immediately halt if the error is a unrecoverable failure.
-			return
-		}
 
-		if txnState.NumNestedTransactions() > 1 {
-			if processErr == nil {
-				// This is a fvm internal programming error.  We forgot to
-				// call Commit somewhere in the control flow.  We should halt.
-				processErr = fmt.Errorf(
-					"successfully executed transaction has unexpected " +
-						"nested transactions.")
-				return
-			} else {
-				restartErr := txnState.RestartNestedTransaction(nestedTxnId)
-				if restartErr != nil {
-					// This should never happen since merging views should
-					// never fail.
-					processErr = fmt.Errorf(
-						"cannot restart nested transaction on error: %w "+
-							"(original processErr: %v)",
-						restartErr,
-						processErr)
-					return
-				}
-
-				ctx.Logger.Warn().Msg(
-					"transaction had unexpected nested transactions, " +
-						"which have been restarted.")
-
-				// Note: proc.Err is set by TransactionProcedure.
-				proc.Logs = make([]string, 0)
-				proc.Events = make([]flow.Event, 0)
-				proc.ServiceEvents = make([]flow.Event, 0)
-			}
-		}
-
-		// based on the contract and frozen account updates we decide how to
-		// clean up the programs for failed transactions we also do the same as
-		// transaction without any deployed contracts
-		programs.AddInvalidator(modifiedSets)
-
-		commitErr := txnState.Commit(nestedTxnId)
-		if commitErr != nil {
-			processErr = fmt.Errorf(
-				"transaction invocation failed when merging state: %w "+
-					"(original processErr: %v)",
-				commitErr,
-				processErr)
-		}
-	}()
-
-	mergeErrorShouldEarlyExit := func(err error) bool {
-		if err == nil {
-			return false
-		}
-
-		if errors.IsFailure(err) {
-			if processErr == nil {
-				processErr = err
-			} else {
-				processErr = fmt.Errorf(
-					"%w (collected errors: %v)",
-					err,
-					processErr)
-			}
-			return true
-		}
-
-		processErr = multierror.Append(processErr, err)
-		return false
-	}
+	errs := errors.NewErrorsCollector()
 
 	env := NewTransactionEnvironment(ctx, txnState, programs, proc.Transaction, proc.TxIndex, span)
 
 	var txError error
 	modifiedSets, txError = i.normalExecution(proc, txnState, env)
-	if mergeErrorShouldEarlyExit(txError) {
-		return
+	if errs.CollectShouldEarlyExit(txError) {
```

Comment:
Just a minor suggestion:

Not the biggest fan of the name: `CollectShouldEarlyExit`. Since it is no longer part of the invoker, the `ShouldEarlyExit` becomes a bit un-intuitive what should be exited.

Perhaps this could be in this form: `if errs.Collect(txError).CollectedFailure() {`?

Reply:
it's a reasonable suggestion.  i'll rename HasError to CollectedError as well